src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx

```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
+      if (i === 26) return node[' '];
+      else if (i >= 0 && i <= 25) {
+        return node[String.fromCharCode(i + 97)];
+      }
+      return null;
+    }
+    const letters = [
+      'a',
+      'b',
+      'c',
+      'd',
+      'e',
+      'f',
+      'g',
+      'h',
+      'i',
+      'j',
+      'k',
+      'l',
+      'm',
+      'n',
+      'o',
+      'p',
+      'q',
+      'r',
+      's',
+      't',
+      'u',
+      'v',
+      'w',
+      'x',
+      'y',
+      'z',
+      ' '
+    ];
+
+    function toEnd(query: any) {
+      let node = trie;
+      for (let i = 0; i < query.length; i++) {
+        const char = query[i];
+        if (node[char]) {
+          node = node[char];
+        } else {
+          return null;
+        }
+      }
+      return node;
+    }
+
+    function recur(node: any, path: string) {
+      if (node == null || limit < 0) {
+        return;
+      }
+      if (node.value) {
+        ans.push(path);
+        limit--;
+      }
+      for (let i = 0; i < 27; i++) {
+        recur(next(node, i), path + letters[i]);
+      }
+    }
+    const ans = [] as any[];
+    recur(toEnd(str), str);
+    return ans;
+  }
+
+  const voidSearch = (str: string) => {
+    function lineMap(array: any[]) {
+      if (array == null || array[0] == null || array[1] == null || array[2] == null) {
+        return { title: '', url: '' };
+      }
+      return {
+        title: textbook[array[0]][array[1]][array[2]],
+        url: SearchUrl + array[0] + array[1]
+      };
+    }
+    const SearchUrl = '..';
+    let words = str.toLowerCase().split(' ');
+    words = words.filter(word => word !== '');
+    if (searchQuery.length === 0 || words.length === 0) {
+      setQueryResult([]);
+      return;
+    }
+    let globalAns = queryTextTrie(words[0]).map((array: any) => lineMap(array));
+
+    globalAns = globalAns.filter((obj: any) => obj.title.toLowerCase().includes(str.toLowerCase()));
+    return globalAns;
+  };
+
+  function sentenceAutoComplete(str: string, limit: number, jsonData: any) {
+    const words = str
+      .toLowerCase()
+      .split(' ')
+      .filter(word => word !== '');
+    if (words.length < 2) {
+      return autoComplete(str, limit, jsonData);
+    }
+    let pre = words[0];
+    for (let i = 1; i < words.length - 1; i++) {
+      pre += ' ' + words[i];
+    }
+    console.log(pre);
+    const lastWord = words[words.length - 1];
+    let preResults = voidSearch(pre);
+    if (preResults == null) {
+      return [];
+    }
+    preResults = preResults.filter((obj: any) => obj.title.includes(lastWord));
+    console.log(preResults);
+    if (preResults.length < 1) {
+      return [];
+    }
+    let lastwords = autoComplete(lastWord, 3000, jsonData);
+    console.log(lastwords);
+    lastwords = lastwords.filter(
+      (word: string) =>
+        preResults.filter((obj: any) => obj.title.toLowerCase().includes(pre + ' ' + word)).length >
+        0
+    );
```

Comment:
What is this code chunk trying to do? `.includes()` gives either `true` or `false`. What's the `.length > 0` for?

Reply:
These are the types of issues that will be caused by having `any` for everything; it's even worse than writing in JavaScript as writing `any` disables all checks by your editor too.

Reply:
Did you mean to just stop after `.includes()`? Can you explain how this entire function works, step by step, in plain English?

Reply:
sorry for bringing this up so late; Maybe you could skip deep "cleaning up" of my code, since many of them are meant to be abolished; I have extended it into CP3108B, and will keep on working on it next semester to properly clean it up

Reply:
As for this function, it goes like: it firstly separate the content that in the input line into two parts, parts that are completed and parts that are not. for example, "As for this fu" will be separated into ""As for this" and "fu". 


Then, "voidSearch" search for all the occurrences of the phrase "As for this" in the textbook coupled with the complete sentence they appear in the textbook. let us say there are only two results "As for this function" "as for this example".

Then, I use the autocomplete to generate guesses for the last word, for example, "fun", "function" "fuel"......

then I check whether they appear in the sentences and filter out absent ones. For example, only "function" appear in the  "As for this function" "as for this example".

so for my sentenceComplete, everyTime it can only produce the guess for the next one word (shrug)